datumaro/plugins/synthetic_data/utils.py

```diff
@@ -0,0 +1,167 @@
+# Copyright (C) 2022 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+import os.path as osp
+import warnings
+
+import cv2 as cv
+import numpy as np
+import requests
+
+warnings.filterwarnings("ignore", message=r"(invalid value|overflow) encountered")
+
+
+class IFSFunction:
+    NUM_PARAMS = 6
+
+    def __init__(self, rng, prev_x, prev_y):
+        self.function = []
+        self.xs, self.ys = [prev_x], [prev_y]
+        self.select_function = []
+        self.cum_proba = 0.0
+        self._rng = rng
+
+    def add_param(self, params, proba, weights=None):
+        if weights is not None:
+            params = list(np.array(params) * np.array(weights))
+
+        self.function.append(params)
+        self.cum_proba += proba
+        self.select_function.append(self.cum_proba)
+
+    def calculate(self, iterations):
+        prev_x, prev_y = self.xs[0], self.ys[0]
+        next_x, next_y = self.xs[0], self.ys[0]
```

Comment:
If `calculate` is called multiple times, then the initial values will always be taken from `xs[0]`, etc., but the new values will be added to the end of the arrays, ignoring all intermediate values. Those semantics don't seem useful. I would suggest either:

1. Initializing from `xs[-1]`, etc. (so that each call to `calculate` continues from where the previous one left off); or
2. Truncating `xs` and `ys` to 1 element on every call to `calculate`; or
3. Making sure `calculate` doesn't get called more than once (e.g. by asserting that `xs` and `ys` only have one element).

Reply:
Changed to initialization from xs[-1]